sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyCastsSuite.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.dsl._
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
+import org.apache.spark.sql.types._
+
+class SimplifyCastsSuite extends PlanTest {
+
+  object Optimize extends RuleExecutor[LogicalPlan] {
+    val batches = Batch("SimplifyCasts", FixedPoint(50), SimplifyCasts) :: Nil
+  }
+
+  test("non-nullable to non-nullable array cast") {
+    val input = LocalRelation('a.array(ArrayType(IntegerType, false)))
+    val plan = input.select('a.cast(ArrayType(IntegerType, false)).as("casted")).analyze
+    val optimized = Optimize.execute(plan)
+    val expected = input.select('a.as("casted")).analyze
+    comparePlans(optimized, expected)
+  }
+
+  test("non-nullable to nullable array cast") {
+    val input = LocalRelation('a.array(ArrayType(IntegerType, false)))
+    val array_intPrimitive = 'a.array(ArrayType(IntegerType, false))
+    val plan = input.select('a.cast(ArrayType(IntegerType, true)).as("casted")).analyze
+    val optimized = Optimize.execute(plan)
+    val expected = input.select('a.as("casted")).analyze
+    comparePlans(optimized, expected)
+  }
+
+  test("nullable to non-nullable array cast") {
+    val input = LocalRelation('a.array(ArrayType(IntegerType, true)))
+    val plan = input.select('a.cast(ArrayType(IntegerType, false)).as("casted")).analyze
+    val optimized = Optimize.execute(plan)
+    comparePlans(optimized, plan)
+  }
+
+  test("nullable to nullable array cast") {
+    val input = LocalRelation('a.array(ArrayType(IntegerType, true)))
+    val plan = input.select('a.cast(ArrayType(IntegerType, true)).as("casted")).analyze
+    val optimized = Optimize.execute(plan)
+    val expected = input.select('a.as("casted")).analyze
+    comparePlans(optimized, expected)
+  }
+
+  def map(keyType: DataType, valueType: DataType, nullable: Boolean): AttributeReference =
```

Comment:
This is because [current map()](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala#L246) cannot pass information on `nullable`.